bnb_yaz/new_bnb_downloader.py

```diff
@@ -0,0 +1,385 @@
+                util.Info("Successfully uploaded MARC file: " + file)
+                # move uploaded MARC file to loaded directory
+                os.rename(file_with_path, working_directory + "loaded/" + file)
+                util.Info("Moved uploaded MARC file to loaded directory: " + working_directory + "loaded/" + file)
+    
+
+# Write summary report
+def WriteSummaryReport() -> str:
+    summary_report = "Detailed Summary Report of BNB Downloader\n"
+    summary_report += "=================================\n\n"
+    for file_name, info in new_bnb_info.items():
+        summary_report += f"File: {file_name}\n"
+        summary_report += f"  Extracted IDs: {len(info['list_of_id_extracted'])}\n"
+        summary_report += f"  Existing IDs: {len(info['list_of_existing_id'])}\n"
+        summary_report += f"  New IDs: {len(info['list_of_new_id'])}\n"
+        summary_report += f"  List of Not Found IDs: {', '.join(info['list_of_not_found_id'])}\n"
+        summary_report += f"  Not Found IDs: {len(info['list_of_not_found_id'])}\n"
+        summary_report += f"  List of Error IDs: {', '.join(info['list_of_error_id'])}\n"
+        summary_report += f"  Error IDs: {len(info['list_of_error_id'])}\n"
+        summary_report += f"  Downloaded IDs: {len(info['list_of_downloaded_id'])}\n\n"
+        
+    summary_report += "BNB Downloader Summary Report\n"
+    summary_report += "=================================\n\n"
+    summary_report += f"List of processed PDF files: {', '.join(pdf_files)}\n"
+    summary_report += f"Total PDF files processed: {len(pdf_files)}\n"
+    summary_report += f"Total BNB IDs extracted: {total_id_extracted}\n"
+    summary_report += f"Total new BNB IDs to download: {len(set_of_new_ids)}\n"
+    summary_report += f"Total existing BNB IDs: {len(set_of_existing_ids)}\n"
+    summary_report += f"Total of not found BNB IDs: {sum(len(info['list_of_not_found_id']) for info in new_bnb_info.values())}\n"
+    summary_report += f"Total of error BNB IDs: {sum(len(info['list_of_error_id']) for info in new_bnb_info.values())}\n"
+    summary_report += f"Total downloaded BNB IDs: {number_of_downloaded_ids}\n\n"
+    
+    
+    return summary_report
+    
+    
+    
+# Main function
+def Main():
+    # check the command line arguments
+    if len(sys.argv) != 3:
+        util.Error("Invalid number of arguments. Usage: new_bnb_downloader.py <email> <working_directory>")
+    
+    # global variables that will be changed
+    global working_directory, email_recipient, db_connection, bnb_yaz_client
+    
+    email_recipient = sys.argv[1]
+    working_directory = sys.argv[2]
+    
+    # set default email recipient in util module, needed for util.SendEmail() when error occurs
+    util.default_email_recipient = email_recipient
+    
+    util.Info("Starting BNB Downloader...")
+    
+    CheckWorkingDirectory()
+    
+    util.Info("Initializing database...")
+    db_connection = InitializeDatabase()
```

Comment:
- Initializing the Database should be the first thing you do, even before scanning the working directory for PDFs
- There should be sub-sections of the code for the 2 parts of processing which group the function calls a bit better

Example:

- Initialize Database Connection
- Read incoming IDs from PDF files
  - CheckWorkingDirectory()
  - ProcessingPDFFiles()
- Download IDs from BNB Server
  - GetNonDeliveredIDs()
  - ConnectToBNBServer()
  - DownloadRecords()